src/app/beer_garden/db/mongo/models.py

```diff
@@ -409,6 +409,10 @@ def pre_serialize(self):
             self.parameters_gridfs = None
 
     def save(self, *args, **kwargs):
+        # Set has_parent if not already set
+        if self.has_parent is None:
```

Comment:
Does it even make sense that `has_parent` is a thing that you can set on the `Request`? As is, and seemingly unchanged here, one could instantiate a Request that sets a `parent`, but then explicitly sets `has_parent` to `False`.  Is that a valid / desired state for the Request to be in?

Would it make more sense to just have the `Request.__init__` set `has_parent` based on the value of `parent`? Or maybe not even write down `has_parent` and just have it be a computed value? There are things like this elsewhere in the code:

```python
# Clear parent before forwarding so the child doesn't freak out about an                                                                                                                                   
# unknown request                                                                                                                                                                                          
operation.model.parent = None                                                                                                                                                                              
operation.model.has_parent = False
```

It feels a bit like having `has_parent` as its own value is just a recipe for it not getting set at some point when its supposed to and thus being out of sync with reality.

Reply:
Yeah, this is pretty gross. The only reason `has_parent` exists at all is because we need to construct some compound indexes in mongo to prevent our request index page from being too slow to be usable. And you can't construct an index from a reference field, or at least you couldn't when we first needed this.

So yeah, there are weird states you can get in that don't make sense (for example, the one this fixes). But I think we need to keep `has_parent` as a "real" database field, unfortunately.

Reply:
Those compound indexes you're referring to look like they could use some love. I suspect that there's some work that could be done there that would obviate the need for this whole `has_parent` thing, but in the interest of not blowing up the scope of your work here, let's carry on under the presumption that the field needs to be written down.

It still seems like we should protect ourselves from setting this incorrectly.  Is the validation that you refer to that happens on non-Admin requests (the `RequestValidator` stuff you altered here) unrelated to the mongoengine document model validation that happens?  It looked like it might be.  If that's the case, I'd suggest:

* This sort of validation and modification seems to belong in the [clean](https://docs.mongoengine.org/guide/validation.html#custom-validation) rather than the `save` method.
* You should raise a `ModelValidationError` if `has_parent` is `False` while `parent` is actually set.

Reply:
Moved into `clean`.

Yep, those indexes are ugly. But they're the only way I could find to have any sort of reasonable performance on the request index page once you start filtering and sorting. We need to support ~750k requests and if mongo has to resort to a COLSCAN it'll take 10s of seconds.